graylog2-server/src/main/java/org/graylog2/utilities/CIDRPatriciaTrie.java

```diff
@@ -0,0 +1,235 @@
+                    shortestV4 = prefixLength;
+                }
+                cleanTrie.put(entry.getKey(), new Node(data.rangeName, data.rangeIsIPv6, data.expireAfter));
+            }
+        }
+        final CIDRPatriciaTrie copy = new CIDRPatriciaTrie();
+        copy.trie = cleanTrie;
+        copy.shortestV4Prefix = shortestV4;
+        copy.shortestV6Prefix = shortestV6;
+        return copy;
+    }
+
+    public void insertCIDR(String cidr, String rangeName) {
+        insertCIDR(cidr, rangeName, 0L);
+    }
+
+    /**
+     * Insert a CIDR range into the trie with a time-to-live
+     *
+     * @param cidr        properly formatted CIDR address (must include '/rangePrefix' even if it is a single address
+     * @param rangeName   the name of the CIDR range
+     * @param expireAfter epoch time in millis after which the CIDR should be expired
+     */
+    public void insertCIDR(String cidr, String rangeName, long expireAfter) {
+        final String[] parts = cidr.split("/");
+        final String ip = parts[0];
+        final int prefixLength;
+        try {
+            prefixLength = Integer.parseInt(parts[1]);
+        } catch (ArrayIndexOutOfBoundsException e) {
+            throw new IllegalArgumentException("Unable to parse invalid CIDR range: " + cidr);
+        }
+
+        // Get binary representation of the IP
+        final String binaryIP = toBinaryString(ip, prefixLength);
+        final boolean isIPV6 = ip.contains(":");
+        final Node node = new Node(rangeName, isIPV6, expireAfter);
+        trie.put(binaryIP, node);
+        if (isIPV6 && (shortestV6Prefix == -1 || prefixLength < shortestV6Prefix)) {
+            this.shortestV6Prefix = prefixLength;
+        } else if (!isIPV6 && (shortestV4Prefix == -1 || prefixLength < shortestV4Prefix)) {
+            this.shortestV4Prefix = prefixLength;
+        }
+    }
+
+    public String longestPrefixRangeLookup(String ip) {
+        return longestPrefixRangeLookupWithTtl(ip, 0L);
+    }
+
+    /**
+     * Returns the rangeName of the range with the longest prefix that contains the IP address or null if one does not
+     * exist.
+     *
+     * @param ip               IP address to check against the collection of ranges
+     * @param lookupTimeMillis time lookup was performed in epoch time milliseconds or 0 if node expiry is not a concern
+     * @return the name of the range with the longest prefix that contains the IP if it exists, null otherwise
+     */
+    public String longestPrefixRangeLookupWithTtl(String ip, long lookupTimeMillis) {
+        if (isEmpty()) {
+            return null;
+        }
+        final String binaryIP = toBinaryString(ip, -1);
+        final boolean lookupIsIPv6 = ip.contains(":");
+
+        final int shortestPrefixForType = lookupIsIPv6 ? shortestV6Prefix : shortestV4Prefix;
+        for (int i = binaryIP.length(); i >= shortestPrefixForType; i--) {
```

Comment:
I've half convinced myself that we can significantly improve the performance of this by using the `headMap` and looking at the last value, but I think I'm thinking about it wrong.
For now we can probably leave it like this, I'll explore some optimizations outside of this PR.
E.g. maybe there are some binary search tricks we can perform on the IP address to limit the number of tree comparisons we have to perform.